File: web/src/pipeline.js
import { inverse, Matrix } from '../vendor/ml-matrix.js';

import { optimize } from './bundle.js';
import { describeKeypoints } from './descriptor.js';
import {
  MIN_POINTS_PNP,
  PNP_SAMPLE_SIZE,
  decomposeEssential,
  ransacFundamental,
  solvePnp,
  triangulatePoint,
} from './estimation.js';
import { identity, normalizePixel, reprojectionError, rotationAngle } from './geometry.js';
import { findKeypoints } from './keypoints.js';
import { buildDifferenceOfGaussian, buildGaussianPyramid, octaveCount } from './pyramid.js';

export const FUNDAMENTAL_THRESHOLD = 0.3;
export const FUNDAMENTAL_ITERATIONS = 1000;
export const LOWE_RATIO = 0.75;
export const PNP_THRESHOLD = 20;
export const PNP_ITERATIONS = 800;
export const BUNDLE_ITERATIONS = 12;
export const MIN_PAIR_MATCHES = 30;
export const MIN_PNP_MATCHES = 8;
export const MIN_SEED_ANGLE = 3;
export const MAX_SEED_ANGLE = 60;
export const SEED_CANDIDATES = 8;
export const DEFAULT_SEED = 7;

export function detectImage(image, { doubleFirst = true } = {}) {
  const octaves = octaveCount(image.width, image.height) - (doubleFirst ? 0 : 1);
  const pyramid = buildGaussianPyramid(image, Math.max(1, octaves), undefined, doubleFirst);
  const keypoints = findKeypoints(buildDifferenceOfGaussian(pyramid));
  return describeKeypoints(pyramid, keypoints).map((point) => ({
    ...point,
    x: doubleFirst ? point.x : point.x * 2,
    y: doubleFirst ? point.y : point.y * 2,
  }));
}

export function packDescriptors(features) {
  const width = features[0]?.descriptor.length ?? 0;
  const packed = new Float32Array(features.length * width);
  features.forEach((feature, i) => packed.set(feature.descriptor, i * width));
  return { packed, width, count: features.length };
}

export function matchFeatures(left, right, ratio = LOWE_RATIO) {
  const a = left.packed ? left : packDescriptors(left);
  const b = right.packed ? right : packDescriptors(right);
  const { width } = a;
  const limit = ratio * ratio;
  const pairs = [];

  for (let i = 0; i < a.count; i += 1) {
    const base = i * width;
    let best = Infinity;
    let second = Infinity;
    let bestIndex = -1;
    for (let j = 0; j < b.count; j += 1) {
      const other = j * width;
      let sum = 0;
      for (let n = 0; n < width; n += 1) {
        const difference = a.packed[base + n] - b.packed[other + n];
        sum += difference * difference;
        if (sum >= second) break;
      }
      if (sum < best) {
        second = best;
        best = sum;
        bestIndex = j;
      } else if (sum < second) {
        second = sum;
      }
    }
    if (best < limit * second) pairs.push([i, bestIndex]);
  }
  return pairs;
}

const key = (left, right) => (left < right ? `${left}:${right}` : `${right}:${left}`);

function samplePnp(points, pixels, k, kInverse, threshold, iterations, seed) {
  let state = seed >>> 0;
  const random = () => {
    state = (state * 1664525 + 1013904223) >>> 0;
    return state / 4294967296;
  };
  const size = Math.min(Math.max(PNP_SAMPLE_SIZE, MIN_POINTS_PNP), points.length);
  let bestCount = -1;
  let best = null;
  for (let round = 0; round < iterations; round += 1) {
    const pool = points.map((_, i) => i);
    for (let i = 0; i < size; i += 1) {
      const j = i + Math.floor(random() * (pool.length - i));
      [pool[i], pool[j]] = [pool[j], pool[i]];
    }
    const chosen = pool.slice(0, size);
    let pose;
    try {
      pose = solvePnp(chosen.map((i) => points[i]), chosen.map((i) => pixels[i]), k, kInverse);
    } catch {
      continue;
    }
    const count = points.filter((point, i) =>
      reprojectionError(k, pose.r, pose.t, point, pixels[i]) < threshold).length;
    if (count > bestCount) {
      bestCount = count;
      best = pose;
    }
  }
  if (!best) throw new Error('pnp found no candidate');
  const inliers = points.map((point, i) =>
    reprojectionError(k, best.r, best.t, point, pixels[i]) < threshold);
  if (inliers.filter(Boolean).length < MIN_POINTS_PNP) return { ...best, inliers };
  const keptPoints = points.filter((_, i) => inliers[i]);
  const keptPixels = pixels.filter((_, i) => inliers[i]);
  try {
    const refined = solvePnp(keptPoints, keptPixels, k, kInverse);
    const after = points.map((point, i) =>
      reprojectionError(k, refined.r, refined.t, point, pixels[i]) < threshold);
    if (after.filter(Boolean).length < inliers.filter(Boolean).length) {
      return { ...best, inliers };
    }
    return { ...refined, inliers: after };
  } catch {
    return { ...best, inliers };
  }
}

export class Session {
  constructor(features, matches, k, seed = DEFAULT_SEED) {
    this.features = features;
    this.matches = matches;
    this.k = k;
    this.kInverse = inverse(new Matrix(k)).to2DArray();
    this.seed = seed;
    this.tracks = new Map();
  }

  pixel(camera, index) {
    const point = this.features.get(camera)[index];
    return [point.x, point.y];
  }

  pairsFor(left, right) {
    const stored = this.matches.get(key(left, right));
    if (!stored) return null;
    return left < right ? stored : stored.map(([a, b]) => [b, a]);
  }

  initialize(left, right) {
    const pairs = this.pairsFor(left, right);
    if (!pairs) throw new Error(`no matches between ${left} and ${right}`);
    const x1 = pairs.map(([a]) => this.pixel(left, a));
    const x2 = pairs.map(([, b]) => this.pixel(right, b));
    const { f, inliers } = ransacFundamental(
      x1, x2, FUNDAMENTAL_THRESHOLD, FUNDAMENTAL_ITERATIONS, this.seed,
    );
    const keptPairs = pairs.filter((_, i) => inliers[i]);
    const rays1 = x1.filter((_, i) => inliers[i]).map((p) => normalizePixel(p, this.kInverse));
    const rays2 = x2.filter((_, i) => inliers[i]).map((p) => normalizePixel(p, this.kInverse));

    const kt = this.k[0].map((_, j) => this.k.map((row) => row[j]));
    const essential = kt.map((row) => row.map((_, j) =>
      row.reduce((sum, value, n) => sum + value * f[n][j], 0)));
    const e = essential.map((row) => row.map((_, j) =>
      row.reduce((sum, value, n) => sum + value * this.k[n][j], 0)));
    const best = decomposeEssential(e, rays1, rays2);

    const p1 = identity().map((row) => [...row, 0]);
    const p2 = best.r.map((row, i) => [...row, best.t[i]]);
    const points = [];
    const observations = [];
    this.tracks = new Map();
    rays1.forEach((ray, row) => {
      if (!best.mask[row]) return;
      const point = triangulatePoint(p1, p2, ray, rays2[row]);
      const index = points.length;
      points.push(point);
      const [a, b] = keptPairs[row];
      this.tracks.set(`${left}:${a}`, index);
      this.tracks.set(`${right}:${b}`, index);
      observations.push({ point: index, camera: left, pixel: this.pixel(left, a) });
      observations.push({ point: index, camera: right, pixel: this.pixel(right, b) });
    });

    return {
      poses: new Map([[left, { r: identity(), t: [0, 0, 0] }], [right, { r: best.r, t: best.t }]]),
      points,
      observations,
      anchor: left,
    };
  }

  correspondences(reconstruction, camera) {
    const found = new Map();
    for (const seen of reconstruction.poses.keys()) {
      const pairs = this.pairsFor(seen, camera);
      if (!pairs) continue;
      for (const [a, b] of pairs) {
        const point = this.tracks.get(`${seen}:${a}`);
        if (point !== undefined) found.set(point, b);
      }
    }
    return found;
  }

  tryRegister(reconstruction, camera, known) {
    const found = known ?? this.correspondences(reconstruction, camera);
    if (found.size < MIN_PNP_MATCHES) {
      return { rejected: { camera, reason: 'too few correspondences', matches: found.size } };
    }
    const indices = [...found.keys()];
    const points = indices.map((i) => reconstruction.points[i]);
    const pixels = indices.map((i) => this.pixel(camera, found.get(i)));

    let pose;
    try {
      pose = samplePnp(points, pixels, this.k, this.kInverse,
        PNP_THRESHOLD, PNP_ITERATIONS, this.seed);
    } catch {
      return { rejected: { camera, reason: 'pnp failed', matches: found.size } };
    }
    const inlierCount = pose.inliers.filter(Boolean).length;
    if (inlierCount < MIN_POINTS_PNP) {
      return {
        rejected: {
          camera, reason: 'too few pnp inliers', matches: found.size, inliers: inlierCount,
        },
      };
    }

    const poses = new Map(reconstruction.poses);
    poses.set(camera, { r: pose.r, t: pose.t });
    const observations = [...reconstruction.observations];
    const added = [];
    indices.forEach((index, n) => {
      if (reprojectionError(this.k, pose.r, pose.t, points[n], pixels[n]) >= PNP_THRESHOLD) return;
      observations.push({ point: index, camera, pixel: pixels[n] });
      added.push(`${camera}:${found.get(index)}`);
    });

    const candidate = {
      poses, points: reconstruction.points.map((p) => [...p]), observations,
      anchor: reconstruction.anchor,
    };
    const result = optimize(candidate, this.k, BUNDLE_ITERATIONS);
    if (result.report.isDegenerate) {
      return {
        rejected: {
          camera,
          reason: 'would make pose underdetermined',
          matches: found.size,
          inliers: inlierCount,
          excessNullDimensions: result.report.excessNullDimensions,
        },
      };
    }
    for (const track of added) this.tracks.set(track, this.tracks.get(track) ?? -1);
    added.forEach((track, n) => this.tracks.set(track, indices[n]));
    return { reconstruction: result.reconstruction, report: result.report };
  }

  scoreSeed(left, right) {
    let reconstruction;
    try {
      reconstruction = this.initialize(left, right);
    } catch {
      return null;
    }
    if (reconstruction.points.length < MIN_PAIR_MATCHES) return null;

    const pose = reconstruction.poses.get(right);
    const angle = rotationAngle(pose.r);
    if (!Number.isFinite(angle)) return null;
    if (angle < MIN_SEED_ANGLE || angle > MAX_SEED_ANGLE) return null;

    const { report } = optimize(reconstruction, this.k, 0);
    if (report.isDegenerate) return null;
    return { pair: [left, right], points: reconstruction.points.length, angle, rms: report.rms };
  }

  chooseSeed() {
    const ranked = [...this.matches.entries()]
      .map(([pairKey, pairs]) => ({ pair: pairKey.split(':').map(Number), count: pairs.length }))
      .sort((a, b) => b.count - a.count)
      .slice(0, SEED_CANDIDATES);
    const scored = [];
    for (const { pair } of ranked) {
      const result = this.scoreSeed(pair[0], pair[1]);
      if (result) scored.push(result);
    }
    if (scored.length === 0) return null;
    scored.sort((a, b) => b.points - a.points);
    return scored[0];
  }

  run(seedPair) {
    let reconstruction = this.initialize(seedPair[0], seedPair[1]);
    let { reconstruction: refined, report } = optimize(reconstruction, this.k, BUNDLE_ITERATIONS);
    reconstruction = refined;

    const rejected = [];
    const history = [{
      camera: seedPair[1], rms: report.rms,
      cameras: reconstruction.poses.size, points: reconstruction.points.length,
    }];
    const blocked = new Set();

    for (;;) {
      const remaining = [...this.features.keys()]
        .filter((c) => !reconstruction.poses.has(c) && !blocked.has(c));
      if (remaining.length === 0) break;
      const shared = new Map(remaining.map((camera) =>
        [camera, this.correspondences(reconstruction, camera)]));
      remaining.sort((a, b) => shared.get(b).size - shared.get(a).size);

      let progressed = false;
      for (const camera of remaining) {
        const outcome = this.tryRegister(reconstruction, camera, shared.get(camera));
        if (outcome.rejected) {
          rejected.push(outcome.rejected);
          blocked.add(camera);
          continue;
        }
        reconstruction = outcome.reconstruction;
        report = outcome.report;
        history.push({
          camera, rms: report.rms,
          cameras: reconstruction.poses.size, points: reconstruction.points.length,
        });
        progressed = true;
        break;
      }
      if (!progressed) break;
    }
    return { reconstruction, report, rejected, history };
  }
}

File: sfm/core/pipeline.py
from dataclasses import dataclass, field

import numpy as np

from sfm.core.bundle import Observation, Reconstruction, optimize
from sfm.core.estimation import (
    MIN_POINTS_PNP,
    decompose_essential,
    ransac_fundamental,
    ransac_pnp,
    triangulate,
)
from sfm.core.geometry import (
    angle_between_directions,
    camera_center,
    normalize_pixels,
    reprojection_error,
)

FUNDAMENTAL_THRESHOLD = 0.3
FUNDAMENTAL_ITERATIONS = 1000
PNP_THRESHOLD = 20.0
PNP_ITERATIONS = 800
MIN_PNP_MATCHES = 8
BUNDLE_ITERATIONS = 12
TRIANGULATION_THRESHOLD = 4.0
MIN_TRIANGULATION_ANGLE = 3.0
GROW_POINTS = False
DEFAULT_SEED = 7


@dataclass
class Rejection:
    camera: int
    reason: str
    matches: int = 0
    inliers: int = 0
    excess_null_dimensions: int = 0


@dataclass
class PipelineResult:
    reconstruction: Reconstruction
    report: object
    rejected: list = field(default_factory=list)
    history: list = field(default_factory=list)


class Session:
    def __init__(self, features, matches, k, seed=DEFAULT_SEED, grow_points=GROW_POINTS):
        self.features = features
        self.matches = matches
        self.k = k
        self.k_inv = np.linalg.inv(k)
        self.seed = seed
        self.grow_points = grow_points

    def pixels(self, camera, indices):
        return self.features[camera].keypoints[indices]

    def pair(self, left, right):
        key = (left, right) if left < right else (right, left)
        pairs = self.matches[key]
        return pairs if left < key[1] or left == key[0] else pairs[:, ::-1]

    def initialize(self, left, right):
        pairs = self.matches[(left, right)]
        x1 = self.pixels(left, pairs[:, 0])
        x2 = self.pixels(right, pairs[:, 1])
        f, inliers = ransac_fundamental(x1, x2, FUNDAMENTAL_THRESHOLD,
                                        FUNDAMENTAL_ITERATIONS, self.seed)
        pairs = pairs[inliers]
        rays1 = normalize_pixels(x1[inliers], self.k_inv)
        rays2 = normalize_pixels(x2[inliers], self.k_inv)
        r, t, mask, _ = decompose_essential(self.k.T @ f @ self.k, rays1, rays2)

        p1 = np.hstack([np.eye(3), np.zeros((3, 1))])
        points = triangulate(p1, np.hstack([r, t.reshape(3, 1)]), rays1, rays2)
        poses = {left: (np.eye(3), np.zeros(3)), right: (r, t)}
        observations, kept, tracks = [], [], {}
        for row, (pair, good) in enumerate(zip(pairs, mask)):
            if not good:
                continue
            index = len(kept)
            kept.append(points[row])
            tracks[(left, pair[0])] = index
            tracks[(right, pair[1])] = index
            observations.append(Observation(index, left, self.pixels(left, pair[0])))
            observations.append(Observation(index, right, self.pixels(right, pair[1])))
        recon = Reconstruction(poses, np.array(kept), observations, anchor=left)
        self.tracks = tracks
        return recon

    def _correspondences(self, recon, camera):
        found = {}
        for (seen_camera, feature), point in self.tracks.items():
            if seen_camera not in recon.poses:
                continue
            key = (seen_camera, camera) if seen_camera < camera else (camera, seen_camera)
            pairs = self.matches.get(key)
            if pairs is None:
                continue
            if seen_camera < camera:
                hits = pairs[pairs[:, 0] == feature][:, 1]
            else:
                hits = pairs[pairs[:, 1] == feature][:, 0]
            for hit in hits:
                found[point] = int(hit)
        return found

    def _try_register(self, recon, camera):
        found = self._correspondences(recon, camera)
        if len(found) < MIN_PNP_MATCHES:
            return None, Rejection(camera, "too few correspondences", len(found))

        indices = list(found)
        points = recon.points[indices]
        pixels = self.pixels(camera, [found[i] for i in indices])
        try:
            r, t, inliers = ransac_pnp(points, pixels, self.k, PNP_THRESHOLD,
                                       PNP_ITERATIONS, self.seed)
        except (RuntimeError, ValueError):
            return None, Rejection(camera, "pnp failed", len(found))
        if inliers.sum() < MIN_POINTS_PNP:
            return None, Rejection(camera, "too few pnp inliers", len(found), int(inliers.sum()))

        poses = dict(recon.poses)
        poses[camera] = (r, t)
        observations = list(recon.observations)
        errors = reprojection_error(self.k, r, t, points, pixels)
        for index, error in zip(indices, errors):
            if error < PNP_THRESHOLD:
                observations.append(Observation(index, camera, self.pixels(camera, found[index])))
                self.tracks[(camera, found[index])] = index
        candidate = Reconstruction(poses, recon.points.copy(), observations, recon.anchor)
        refined, report = optimize(candidate, self.k, BUNDLE_ITERATIONS)
        if report.is_degenerate:
            for index in indices:
                self.tracks.pop((camera, found[index]), None)
            return None, Rejection(camera, "would make pose underdetermined", len(found),
                                   int(inliers.sum()), report.excess_null_dimensions)

        if self.grow_points:
            points, observations = self._grow(refined, camera)
            grown = Reconstruction(refined.poses, points, observations, refined.anchor)
            refined, report = optimize(grown, self.k, BUNDLE_ITERATIONS)
        return (refined, report), Rejection(camera, "", len(found), int(inliers.sum()))

    def _grow(self, recon, camera):
        points = list(recon.points)
        observations = list(recon.observations)
        r_new, t_new = recon.poses[camera]
        p_new = np.hstack([r_new, t_new.reshape(3, 1)])
        for other in recon.poses:
            if other == camera:
                continue
            key = (other, camera) if other < camera else (camera, other)
            pairs = self.matches.get(key)
            if pairs is None:
                continue
            r_old, t_old = recon.poses[other]
            p_old = np.hstack([r_old, t_old.reshape(3, 1)])
            for row in pairs:
                a, b = (row[0], row[1]) if other < camera else (row[1], row[0])
                if (other, a) in self.tracks or (camera, b) in self.tracks:
                    continue
                pixel_old = self.pixels(other, a)
                pixel_new = self.pixels(camera, b)
                ray_old = normalize_pixels(pixel_old[None, :], self.k_inv)[0]
                ray_new = normalize_pixels(pixel_new[None, :], self.k_inv)[0]
                point = triangulate(p_old, p_new, ray_old[None, :], ray_new[None, :])[0]
                if (r_old @ point + t_old)[2] <= 0 or (r_new @ point + t_new)[2] <= 0:
                    continue
                to_old = camera_center(r_old, t_old) - point
                to_new = camera_center(r_new, t_new) - point
                if angle_between_directions(to_old, to_new) < MIN_TRIANGULATION_ANGLE:
                    continue
                errors = [reprojection_error(self.k, r_old, t_old, point[None, :], pixel_old[None, :])[0],
                          reprojection_error(self.k, r_new, t_new, point[None, :], pixel_new[None, :])[0]]
                if max(errors) > TRIANGULATION_THRESHOLD:
                    continue
                index = len(points)
                points.append(point)
                self.tracks[(other, a)] = index
                self.tracks[(camera, b)] = index
                observations.append(Observation(index, other, pixel_old))
                observations.append(Observation(index, camera, pixel_new))
        return np.array(points), observations

    def run(self, seed_pair):
        recon = self.initialize(*seed_pair)
        recon, report = optimize(recon, self.k, BUNDLE_ITERATIONS)
        result = PipelineResult(recon, report)
        result.history.append((seed_pair[1], report.rms, len(recon.poses), len(recon.points)))

        blocked = set()
        while True:
            remaining = [c for c in self.features
                         if c not in result.reconstruction.poses and c not in blocked]
            if not remaining:
                break
            scored = sorted(remaining,
                            key=lambda c: -len(self._correspondences(result.reconstruction, c)))
            progressed = False
            for camera in scored:
                outcome, rejection = self._try_register(result.reconstruction, camera)
                if outcome is None:
                    result.rejected.append(rejection)
                    blocked.add(camera)
                    continue
                result.reconstruction, result.report = outcome
                result.history.append((camera, result.report.rms,
                                       len(result.reconstruction.poses),
                                       len(result.reconstruction.points)))
                progressed = True
                break
            if not progressed:
                break
        return result